hyper-capi/include/hyper.h

```diff
@@ -0,0 +1,323 @@
+
+// Set the write function for this IO transport.
+//
+// Data from the `buf` pointer should be written to the transport, up to
+// `buf_len` bytes. The number of bytes written should be the return value.
+//
+// If no data can currently be written, the `waker` should be cloned and
+// registered with whatever polling mechanism is used to signal when data
+// is available later on. The return value should be `HYPER_IO_PENDING`.
+//
+// If there is an irrecoverable error reading data, then `HYPER_IO_ERROR`
+// should be the return value.
+void hyper_io_set_write(hyper_io *io, size_t (*func)(void *userdata, hyper_context *ctx, const uint8_t *buf, size_t buf_len));
+
+// HTTP Requests
+
+// Construct a new HTTP request.
+hyper_request *hyper_request_new(void);
+
+// Free an HTTP request if not going to send it on a client.
+void hyper_request_free(hyper_request *request);
+
+// Set the HTTP Method of the request.
+hyper_code hyper_request_set_method(hyper_request *request, uint8_t *method, size_t method_len);
+
+// Set the URI of the request.
+hyper_code hyper_request_set_uri(hyper_request *request, uint8_t *uri, size_t uri_len);
+
+
+// Gets a reference to the HTTP headers of this request
+//
+// This is not an owned reference, so it should not be accessed after the
+// `hyper_request` has been consumed.
+hyper_headers *hyper_request_headers(hyper_request *request);
+
+
+// HTTP Responses
+
+// Free an HTTP response after using it.
+void hyper_response_free(hyper_response *response);
+
+// Get the HTTP-Status code of this response.
+//
+// It will always be within the range of 100-599.
+uint16_t hyper_response_status(hyper_response *response);
+
+// Gets a reference to the HTTP headers of this response.
+//
+// This is not an owned reference, so it should not be accessed after the
+// `hyper_response` has been freed.
+hyper_headers *hyper_response_headers(hyper_response *response);
+
+// Take ownership of the body of this response.
+//
+// It is safe to free the response even after taking ownership of its body.
+hyper_body *hyper_response_body(hyper_response *response);
+
+// HTTP Headers
+
+// Sets the header with the provided name to the provided value.
+//
+// This overwrites any previous value set for the header.
+void hyper_headers_set(hyper_headers *headers,
+                       const uint8_t* name,
+                       size_t name_len,
+                       const uint8_t *value,
+                       size_t value_len);
+
+// Adds the provided value to the list of the provided name.
+//
+// If there were already existing values for the name, this will append the
+// new value to the internal list.
+void hyper_headers_add(hyper_headers *headers,
+                       const uint8_t* name,
+                       size_t name_len,
+                       const uint8_t *value,
+                       size_t value_len);
```

Comment:
Both `hyper_headers_set` and `hyper_headers_add` return `void` here. What happens when they can't allocate memory for adding the headers to the internal lists? Surely they must be able to return error?

Reply:
This was a mistake in the header file, they did indeed have a `hyper_code` return type in the Rust code, and I've updated the header file.

I've been searching this morning to see if there's a tool that validates that functions defined in header files exist in a compiled library, and so far haven't found one.

Reply:
> I've been searching this morning to see if there's a tool that validates that functions defined in header files exist in a compiled library, and so far haven't found one.

I'm not aware of one either, but `cbindgen` could create the header file automatically for `pub extern "C" fn`s if you like one of the C API styles it supports. That also prevents implementation and header to get out of sync.

Reply:
I am trying out `cbindgen` right now, if all works this could be a good way to prevent these mistakes.